Mypy/Typing check

Current master doesn't pass mypy check:

```
(sgkit-dev) ➜  sgkit git:(master) ✗ mypy --strict .
setup.py:2: error: Skipping analyzing 'setuptools': found module but no type hints or library stubs
sgkit/utils.py:4: error: Function is missing a type annotation
sgkit/tests/test_utils.py:2: error: Skipping analyzing 'pytest': found module but no type hints or library stubs
sgkit/tests/test_utils.py:2: note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
sgkit/tests/test_utils.py:9: error: Call to untyped function "check_array_like" in typed context
sgkit/tests/test_utils.py:12: error: Call to untyped function "check_array_like" in typed context
sgkit/tests/test_utils.py:14: error: Call to untyped function "check_array_like" in typed context
sgkit/tests/test_utils.py:16: error: Call to untyped function "check_array_like" in typed context
sgkit/tests/test_utils.py:18: error: Call to untyped function "check_array_like" in typed context
sgkit/tests/test_utils.py:20: error: Call to untyped function "check_array_like" in typed context
sgkit/api.py:55: error: Call to untyped function "check_array_like" in typed context
sgkit/api.py:56: error: Call to untyped function "check_array_like" in typed context
sgkit/api.py:57: error: Call to untyped function "check_array_like" in typed context
sgkit/api.py:58: error: Call to untyped function "check_array_like" in typed context
sgkit/api.py:59: error: Call to untyped function "check_array_like" in typed context
sgkit/api.py:72: error: Call to untyped function "check_array_like" in typed context
sgkit/api.py:73: error: Unsupported target for indexed assignment ("Mapping[Hashable, Any]")
sgkit/api.py:78: error: Call to untyped function "check_array_like" in typed context
sgkit/api.py:79: error: Unsupported target for indexed assignment ("Mapping[Hashable, Any]")
sgkit/api.py:81: error: Argument "attrs" to "Dataset" has incompatible type "Dict[str, List[str]]"; expected "Mapping[Hashable, Any]"
Found 19 errors in 4 files (checked 7 source files)
```

It goes without saying that typing provides huge benefits (especially over time), there are also many typing limitations in the python (data) ecosystem. Currently our build and pre-commit don't run mypy check, there isn't much code in this repo, but there is a bunch coming in soon. So some questions about this:
 * do we want to enable mypy check for both build and pre-commit right now (I'm +1)
 * we likely won't be able to type everything, so we need to agree which parts need to be typed and which part can be omitted

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -19,6 +19,9 @@ jobs:
         pip install -r requirements.txt -r requirements-dev.txt
     - name: Run pre-commit
       uses: pre-commit/action@v2.0.0
+    - name: Check types with mypy
+      run: |
+        mypy --strict .
     - name: Test with pytest and coverage
       run: |
         pytest -v --cov=sgkit

--- a/README.md
+++ b/README.md
@@ -34,3 +34,5 @@ You can use [pre-commit](https://pre-commit.com/) to check or enforce the coding
 ```bash
 pre-commit install
 ```
+
+Note that pre-commit does not run the `mypy` check, so you will need to run that manually before submitting a pull request.

--- a/setup.cfg
+++ b/setup.cfg
@@ -59,6 +59,9 @@ line_length = 88
 
 [mypy-numpy.*]
 ignore_missing_imports = True
-
+[mypy-pytest.*]
+ignore_missing_imports = True
+[mypy-setuptools]
+ignore_missing_imports = True
 [mypy-sgkit.tests.*]
 disallow_untyped_defs = False

--- a/sgkit/api.py
+++ b/sgkit/api.py
@@ -1,4 +1,4 @@
-from typing import Any, Hashable, List, Mapping
+from typing import Any, Dict, Hashable, List
 
 import xarray as xr
 
@@ -57,7 +57,7 @@ def create_genotype_call_dataset(
     check_array_like(variant_alleles, kind="S", ndim=2)
     check_array_like(sample_id, kind="U", ndim=1)
     check_array_like(call_genotype, kind="i", ndim=3)
-    data_vars: Mapping[Hashable, Any] = {
+    data_vars: Dict[Hashable, Any] = {
         "variant/contig": ([DIM_VARIANT], variant_contig),
         "variant/position": ([DIM_VARIANT], variant_position),
         "variant/alleles": ([DIM_VARIANT, DIM_ALLELE], variant_alleles),
@@ -77,5 +77,5 @@ def create_genotype_call_dataset(
     if variant_id is not None:
         check_array_like(variant_id, kind="U", ndim=1)
         data_vars["variant/id"] = ([DIM_VARIANT], variant_id)
-    attrs = {"contigs": variant_contig_names}
+    attrs: Dict[Hashable, Any] = {"contigs": variant_contig_names}
     return xr.Dataset(data_vars=data_vars, attrs=attrs)

--- a/sgkit/typing.py
+++ b/sgkit/typing.py
@@ -1,0 +1,3 @@
+from typing import Any
+
+DType = Any

--- a/sgkit/utils.py
+++ b/sgkit/utils.py
@@ -5,7 +5,7 @@ import numpy as np
 
 def check_array_like(
     a: Any,
-    dtype: Union[None, str, Set[str]] = None,
+    dtype: Any = None,
     kind: Union[None, str, Set[str]] = None,
     ndim: Union[None, int, Set[int]] = None,
 ) -> None:

--- a/sgkit/utils.py
+++ b/sgkit/utils.py
@@ -2,10 +2,12 @@ from typing import Any, Set, Union
 
 import numpy as np
 
+from .typing import DType
+
 
 def check_array_like(
     a: Any,
-    dtype: Any = None,
+    dtype: DType = None,
     kind: Union[None, str, Set[str]] = None,
     ndim: Union[None, int, Set[int]] = None,
 ) -> None:

--- a/sgkit/utils.py
+++ b/sgkit/utils.py
@@ -1,7 +1,14 @@
+from typing import Any, Set, Union
+
 import numpy as np
 
 
-def check_array_like(a, dtype=None, kind=None, ndim=None):
+def check_array_like(
+    a: Any,
+    dtype: Union[None, str, Set[str]] = None,
+    kind: Union[None, str, Set[str]] = None,
+    ndim: Union[None, int, Set[int]] = None,
+) -> None:
     array_attrs = "ndim", "dtype", "shape"
     for k in array_attrs:
         if not hasattr(a, k):

